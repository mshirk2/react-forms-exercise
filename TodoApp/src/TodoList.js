import React, { useState } from "react";
import { v4 as uuid } from "uuid";
import NewTodoForm from "./NewTodoForm"
import Todo from "./Todo";

const TodoList = () => {
    const [todos, setTodos] = useState([]);

    const addTodo = todo => {
        let newTodo = {...todo, id:uuid()};
        setTodos(todos => [...todos, newTodo]);
    };

    const removeTodo = id => {
        setTodos(todos => todos.filter(todo => todo.id !==id));
    }

    const renderTodos = () => {
        return (
            <div>
                <ul>
                    {todos.map(({id, todoText}) => 
                        <li key={id}>
                            <Todo 
                            id={id}
                            todoText={todoText}
                            handleRemove={removeTodo}
                            />
                        </li>
                    )}
                </ul>
            </div>
        );
    };

    return (
        <div className="TodoList">
            <NewTodoForm addTodo={addTodo} />
            {renderTodos()}
        </div>
    );
};

export default TodoList;