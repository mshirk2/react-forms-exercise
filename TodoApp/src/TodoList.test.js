import React from "react";
import { render, fireEvent } from "@testing-library/react";
import TodoList from "./TodoList";

function addTodo(todoList, task='make working app') {

    fireEvent.change(
        todoList.getByLabelText("To-do"), 
        {target: { value: task}}
    );

    const submitButton = todoList.getByText("Add");
    fireEvent.click(submitButton);
}

it("renders without crashing", function () {
    render(<TodoList />);
});

it("matches snapshot", function () {
    const { asFragment } = render(<TodoList />);
    expect(asFragment()).toMatchSnapshot();
});

it("starts with no todos", function () {
    const todoList = render(<TodoList />);

    expect(todoList.queryByText("X")).toBeFalsy();
});

it("adds new todo", function () {
    const todoList = render(<TodoList />);
    addTodo(todoList);
    const removeButton = todoList.getByText("X");

    expect(removeButton).toBeTruthy();
    expect(removeButton.previousSibling).toContain('make working app');
});

it("clears inputs after adding todo", function () {
    const todoList = render(<TodoList />);
    addTodo(todoList);
    const removeButton = todoList.getByText("X");

    expect(removeButton).toBeTruthy();
    expect(todoList.getByLabelText("To-do").value = "");
});

it("removes todo", function () {
    const todoList = render(<TodoList />);
    addTodo(todoList);
    const removeButton = todoList.getByText("X");
    fireEvent.click(removeButton);

    expect(todoList.queryByText("X")).toBeFalsy();
});