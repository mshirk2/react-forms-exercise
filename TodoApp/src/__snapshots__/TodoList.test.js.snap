// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="TodoList"
  >
    <form>
      <label
        for="todoText"
      >
        To-do
      </label>
      <input
        id="todoText"
        name="todoText"
        type="text"
        value=""
      />
      <button>
        Add
      </button>
    </form>
    <div>
      <ul />
    </div>
  </div>
</DocumentFragment>
`;
